sdk/ton/executor.go

```diff
@@ -0,0 +1,228 @@
+package ton
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand/v2"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	chain_selectors "github.com/smartcontractkit/chain-selectors"
+	commonton "github.com/smartcontractkit/chainlink-ton/pkg/ccip/bindings/common"
+
+	"github.com/smartcontractkit/mcms/sdk"
+	"github.com/smartcontractkit/mcms/types"
+
+	"github.com/xssnick/tonutils-go/ton"
+	"github.com/xssnick/tonutils-go/ton/wallet"
+
+	"github.com/xssnick/tonutils-go/address"
+	"github.com/xssnick/tonutils-go/tlb"
+
+	"github.com/smartcontractkit/chainlink-ton/pkg/bindings/mcms/mcms"
+	"github.com/smartcontractkit/chainlink-ton/pkg/ccip/bindings/ocr"
+)
+
+// sdk.Executor implementation for TON chains, allowing for the execution of operations on the MCMS contract
+type executor struct {
+	sdk.Encoder
+	sdk.Inspector
+
+	wallet *wallet.Wallet
+
+	// Transaction opts
+	amount tlb.Coins
+}
+
+// NewExecutor creates a new Executor for TON chains
+func NewExecutor(encoder sdk.Encoder, client ton.APIClientWrapped, wallet *wallet.Wallet, amount tlb.Coins) (sdk.Executor, error) {
+	if IsNil(encoder) {
+		return nil, errors.New("failed to create sdk.Executor - encoder (sdk.Encoder) is nil")
+	}
+
+	if IsNil(client) {
+		return nil, errors.New("failed to create sdk.Executor - client (ton.APIClientWrapped) is nil")
+	}
+
+	if wallet == nil {
+		return nil, errors.New("failed to create sdk.Executor - wallet (*wallet.Wallet) is nil")
+	}
+
+	return &executor{
+		Encoder:   encoder,
+		Inspector: NewInspector(client, NewConfigTransformer()),
+		wallet:    wallet,
+		amount:    amount,
+	}, nil
+}
+
+func (e *executor) ExecuteOperation(
+	ctx context.Context,
+	metadata types.ChainMetadata,
+	nonce uint32,
+	proof []common.Hash,
+	op types.Operation,
+) (types.TransactionResult, error) {
+	oe, ok := e.Encoder.(OperationEncoder[mcms.Op])
+	if !ok {
+		return types.TransactionResult{}, fmt.Errorf("failed to assert OperationEncoder")
+	}
+
+	bindOp, err := oe.ToOperation(nonce, metadata, op)
+	if err != nil {
+		return types.TransactionResult{}, fmt.Errorf("failed to convert to operation: %w", err)
+	}
+
+	// Encode proofs
+	pe, ok := e.Encoder.(ProofEncoder[mcms.Proof])
+	if !ok {
+		return types.TransactionResult{}, fmt.Errorf("failed to assert ProofEncoder")
+	}
+
+	bindProof, err := pe.ToProof(proof)
+	if err != nil {
+		return types.TransactionResult{}, fmt.Errorf("failed to encode proof: %w", err)
+	}
+
+	body, err := tlb.ToCell(mcms.Execute{
+		QueryID: rand.Uint64(),
+
+		Op:    bindOp,
+		Proof: commonton.SnakeData[mcms.Proof](bindProof),
+	})
+	if err != nil {
+		return types.TransactionResult{}, fmt.Errorf("failed to encode ExecuteBatch body: %w", err)
+	}
+
+	// Map to Ton Address type
+	dstAddr, err := address.ParseAddr(metadata.MCMAddress)
+	if err != nil {
+		return types.TransactionResult{}, fmt.Errorf("invalid timelock address: %w", err)
+	}
+
+	msg := &wallet.Message{
+		Mode: wallet.PayGasSeparately,
```

Comment:
You should also add wallet.IgnoreErrors, TON wallet V5 now enforcing this [flag](https://github.com/tonkeeper/w5/blob/c5465be2312b4fd83cb429904828c771db6f7721/contracts/wallet_v5.fc#L10) to prevent replay attack